probe/ebpf/flow.c

```diff
@@ -357,6 +366,27 @@ int bpf_flow_table(struct __sk_buff *skb)
 			__sync_fetch_and_add(&prev->rtt, tm - prev->start);
 			prev->_flags = RTT_SET;
 		}
+		if(prev->transport_layer.port_src==flow.transport_layer.port_src){
```

Comment:
maybe it would be better to add all the tcp metrics related code in the update_metrics function ? you will have the ab/ba order and the timestamp is passed.

Reply:
3) Same

Reply:
Here I gave it some more thought and I don't think now that this statement is quite accurate. First the ab/ba order relates to the link layer while here we are concerned about the transport layer. Isn't this right?
+ I think there might be a bug in the current implementation so if you could help me clear a few things up I'd really appreciate it. Updatemetrics relies on the identities of the source and the destination in the link layer in order to determine the direction of the flow. but isn't this something that changes very frequently?
I mean if looking up a flow in the flow table relies on a 5-tuple representing the flow then the source and the destination in the link layer play no role in it. furthermore, if, say, A sends packets to B using different paths each time then this comparison in the link level would return false and the metrics would be updated incorrectly. Shouldn't we be using IPs instead?

Reply:
The link layer is there in order to avoid conflict due to overlapping ips(tunneling). That said the ebpf implementation is really a naive version of the gopacket one. Tunneling is not supported and some other things. 

For the IPs vs Link there is a mode to select what are the fields to take into account. 
https://github.com/skydive-project/skydive/blob/master/flow/table.go#L56

I think we should pass the mode to the kernel side using this map https://github.com/skydive-project/skydive/blob/master/flow/probes/ebpf.go#L255